## Find the Runner-Up Score!


In [16]:
if __name__ == '__main__':
    n = int(input())
    if n<2 :
        n=2
    if n>10:
        n=10
    arr = list(map(int, input().split()))
    arr = arr[:n]
    for i in range(len(arr)):
        if arr[i] < -100:
            arr[i] = -100
        if arr[i] > 100:
            arr[i] = 100
#     print(arr)
#     print(max(arr))
    max_num = max(arr)
    while max_num in arr:
        arr.remove(max(arr))
#     print(arr)
    print(sorted(arr, reverse=True)[0])

5
2 3 6 6 5
5


## Finding the percentage

In [13]:
if __name__ == '__main__':
    n = int(input())
    student_marks = {}
    for _ in range(n):
        name, *line = input().split()
        scores = list(map(float, line))
        student_marks[name] = scores
    query_name = input()
    print('{:0.2f}'.format(sum(student_marks[query_name])/len(student_marks[query_name])))

2
a 1 2 3
b 2 3 4
a
2.00


## Detect Floating Point Number

In [19]:
if __name__ == '__main__':
    import re
    t = int((input()))
    for i in range(t):
        n = input()
        print(bool(re.match(r'^[-+]?[0-9]*\.[0-9]+$', n)))

4
1.2
True
1.
False
5..ii
False
r
False


## Re.split()

In [23]:
regex_pattern = r"[.,]+"	# Do not delete 'r'.

import re
print("\n".join(re.split(regex_pattern, input())))

100,000,000
100
000
000


## Lists

In [34]:
if __name__ == '__main__':
    N = int(input())
    number_list = []
    for _ in range(N):
        command, *line = input().split()
        index_number = list(map(int, line))
        if command == 'print':
            print(number_list)
        elif command == 'insert':
            number_list.insert(index_number[0], index_number[1])
        elif command == 'remove':
            number_list.remove(index_number[0])
        elif command == 'append':
            number_list.append(index_number[0])
        elif command == 'sort':
            number_list.sort()
        elif command == 'pop':
            number_list.pop()
        elif command == 'reverse':
            number_list.reverse()

12
insert 0 5
insert 1 10
insert 0 6
print
[6, 5, 10]
remove 6
append 9
append 1
sort
print
[1, 5, 9, 10]
pop
reverse
print
[9, 5, 1]


## Tuples

In [39]:
if __name__ == '__main__':
    n = int(input())
    integer_list = map(int, input().split())
    t = tuple(integer_list)
    print(hash(t))

2
1 2
-3550055125485641917


## Nested Lists

In [27]:
if __name__ == '__main__':
    student_all = []
    n = int(input())
    for _ in range(n):
        name = input()
        score = float(input())
        student_all.append([name, score])
        order = sorted(student_all, key = lambda x: float(x[1]))
#     print(order)
    for i in range(n):
        if order[i][1] != order[0][1]:
            second_low = order[i][1]
            break
    name_result = []
    for i in range(n):
        if order[i][1] == second_low:
            name_result.append(order[i][0])
    for name_ in sorted(name_result):
        print(name_)

5
H
37.21
B
37.21
T
37.2
A
41
Har
39
[['T', 37.2], ['H', 37.21], ['B', 37.21], ['Har', 39.0], ['A', 41.0]]
B
H


## Set .symmetric_difference() Operation

In [28]:
if __name__ == '__main__':
    student_all = []
    no_eng = int(input())
    eng = set(list(map(int, input().split()))[:no_eng])
    no_fr = int(input())
    fr = set(list(map(int, input().split()))[:no_fr])
    print(len(eng.symmetric_difference(fr)))

9
1 2 3 4 5 6 7 8 9
9
10 1 2 3 11 21 55 6 8
8


# Classes

## Class 2 - Find the Torsional Angle

In [173]:
import math

class Points(object):
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z
        
    def __sub__(self, no):
        return Points(self.x - no.x, self.y - no.y, self.z - no.z)

    def dot(self, no):
        return self.x * no.x + self.y * no.y + self.z * no.z

    def cross(self, no):
        return Points(self.y * no.z - self.z * no.y
                    , self.z * no.x - self.x * no.z
                    , self.x * no.y - self.y * no.x)
        
    def absolute(self):
        return pow((self.x ** 2 + self.y ** 2 + self.z ** 2), 0.5)

if __name__ == '__main__':
    points = list()
    for i in range(4):
        a = list(map(float, input().split()))
        points.append(a)

    a, b, c, d = Points(*points[0]), Points(*points[1]), Points(*points[2]), Points(*points[3])
    x = (b - a).cross(c - b)
    y = (c - b).cross(d - c)
    angle = math.acos(x.dot(y) / (x.absolute() * y.absolute()))

    print("%.2f" % math.degrees(angle))

0 4 5
1 7 6
0 5 9
1 7 2
8.19


In [171]:
import math
class Points(object):
    def __init__(self, x, y, z):
        self.x=x
        self.y=y
        self.z=z

    def __sub__(self, no):
        return  Points((self.x-no.x),(self.y-no.y),(self.z-no.z))

    def dot(self, no):
        return (self.x*no.x)+(self.y*no.y)+(self.z*no.z)

    def cross(self, no):
        return Points((self.y*no.z-self.z*no.y),(self.z*no.x-self.x*no.z),(self.x*no.y-self.y*no.x))
        
    def absolute(self):
        return pow((self.x ** 2 + self.y ** 2 + self.z ** 2), 0.5)
if __name__ == '__main__':
    points = list()
    for i in range(4):
        a = list(map(float, input().split()))
        points.append(a)

    a, b, c, d = Points(*points[0]), Points(*points[1]), Points(*points[2]), Points(*points[3])
    x = (b - a).cross(c - b)
    y = (c - b).cross(d - c)
    angle = math.acos(x.dot(y) / (x.absolute() * y.absolute()))

    print("%.2f" % math.degrees(angle))

0 4 5
1 7 6
0 5 9
1 7 2
8.19


## Classes: Dealing with Complex Numbers

In [1]:
# Classes Dealing with Complex Numbers in python - Hacker Rank Solution
# Python 3

import math

class Complex(object):
# Classes Dealing with Complex Numbers in python - Hacker Rank Solution START
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary
        
    def __add__(self, no):
        real = self.real + no.real
        imaginary = self.imaginary + no.imaginary
        return Complex(real,imaginary)
        
    def __sub__(self, no):
        real = self.real - no.real
        imaginary = self.imaginary - no.imaginary
        return Complex(real,imaginary)
        
    def __mul__(self, no):
        real = self.real * no.real - self.imaginary * no.imaginary
        imaginary = self.real * no.imaginary + self.imaginary * no.real
        return Complex(real,imaginary)

    def __truediv__(self, no):
        x = float(no.real ** 2 + no.imaginary ** 2)
        y = self * Complex(no.real, -no.imaginary)
        real = y.real / x
        imaginary = y.imaginary / x
        return Complex(real, imaginary)

    def mod(self):
        real = math.sqrt(self.real ** 2 + self.imaginary ** 2)
        return Complex(real, 0)
# Classes Dealing with Complex Numbers in python - Hacker Rank Solution END

    def __str__(self):
        if self.imaginary == 0:
            result = "%.2f+0.00i" % (self.real)
        elif self.real == 0:
            if self.imaginary >= 0:
                result = "0.00+%.2fi" % (self.imaginary)
            else:
                result = "0.00-%.2fi" % (abs(self.imaginary))
        elif self.imaginary > 0:
            result = "%.2f+%.2fi" % (self.real, self.imaginary)
        else:
            result = "%.2f-%.2fi" % (self.real, abs(self.imaginary))
        return result

if __name__ == '__main__':
    c = map(float, input().split())
    d = map(float, input().split())
    x = Complex(*c)
    y = Complex(*d)
    print(*map(str, [x+y, x-y, x*y, x/y, x.mod(), y.mod()]), sep='\n')

2 1
5 6
7.00+7.00i
-3.00-5.00i
4.00+17.00i
0.26-0.11i
2.24+0.00i
7.81+0.00i


# Strings

## sWAP cASE

In [2]:
def swap_case(s):
    return s.swapcase()

if __name__ == '__main__':
    s = input()
    result = swap_case(s)
    print(result)

HackerRank.com presents "Pythonist 2".
hACKERrANK.COM PRESENTS "pYTHONIST 2".


## String Split and Join

In [3]:
def split_and_join(line):
    # write your code here
    line = line.split(" ")
    line = "-".join(line)
    return line

if __name__ == '__main__':
    line = input()
    result = split_and_join(line)
    print(result)

this is a string
this-is-a-string


## What's Your Name?

In [5]:
def print_full_name(a, b):
    print("Hello {} {}! You just delved into python.".format(a, b))

if __name__ == '__main__':
    first_name = input()
    last_name = input()
    print_full_name(first_name, last_name)

Ross
Taylor
Hello Ross Taylor! You just delved into python.


## Mutations

In [8]:
def mutate_string(string, position, character):
    return string[:position] + character + string[position+1:]

if __name__ == '__main__':
    s = input()
    i, c = input().split()
    s_new = mutate_string(s, int(i), c)
    print(s_new)

abracadabra
5 k
abrackdabra


## Find a string

In [2]:
def count_substring(string, sub_string):
    count = 0
    i=0
    pos=0
    while i<len(string):
        # print(len(string))
        if string[pos:].find(sub_string) != -1:
            # print('string[i:] : ',string[i:])
            # print('i :', i)
            count = count+1
            i = i + string[pos:].find(sub_string) + len(sub_string)
            pos = i-1
            # print('i :', i)
        else:
            i = i+1
            # print('i :', i)
    return count

if __name__ == '__main__':
    string = input().strip()
    sub_string = input().strip()
    
    count = count_substring(string, sub_string)
    print(count)

ABCDCDC
CDC
2


In [5]:
def count_substring(string, sub_string):
    c=0
    for i in range(len(string)):
        if string[i:].startswith(sub_string):
            c +=1
    return c

if __name__ == '__main__':
    string = input().strip()
    sub_string = input().strip()
    
    count = count_substring(string, sub_string)
    print(count)

ABCDCDC
CDC
2


## String Validators

In [8]:
if __name__ == '__main__':
    s = input()
    isalnum = 0
    isalpha = 0
    isdigit = 0
    islower = 0
    isupper = 0
    for char in s:
        if char.isalnum():
            isalnum = 1
        if char.isalpha():
            isalpha = 1
        if char.isdigit():
            isdigit = 1
        if char.islower():
            islower = 1
        if char.isupper():
            isupper = 1

    print(bool(isalnum))
    print(bool(isalpha))
    print(bool(isdigit))
    print(bool(islower))
    print(bool(isupper))

qA2
True
True
True
True
True


## Text Alignment

In [33]:
#Replace all ______ with rjust, ljust or center. 

thickness = int(input()) #This must be an odd number
c = 'H'

#Top Cone
for i in range(thickness):
    print((c*i).rjust(thickness-1)+c+(c*i).ljust(thickness-1))

#Top Pillars
for i in range(thickness+1):
    print((c*thickness).center(thickness*2)+(c*thickness).center(thickness*6))

#Middle Belt
for i in range((thickness+1)//2):
    print((c*thickness*5).center(thickness*6))    

#Bottom Pillars
for i in range(thickness+1):
    print((c*thickness).center(thickness*2)+(c*thickness).center(thickness*6))    

#Bottom Cone
for i in range(thickness):
    print(((c*(thickness-i-1)).rjust(thickness)+c+(c*(thickness-i-1)).ljust(thickness)).rjust(thickness*6))

5
    H    
   HHH   
  HHHHH  
 HHHHHHH 
HHHHHHHHH
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHHHHHHHHHHHHHHHHHHHHHH   
  HHHHHHHHHHHHHHHHHHHHHHHHH   
  HHHHHHHHHHHHHHHHHHHHHHHHH   
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
  HHHHH               HHHHH             
                    HHHHHHHHH 
                     HHHHHHH  
                      HHHHH   
                       HHH    
                        H     


## Text Wrap

In [39]:
import textwrap

def wrap(string, max_width):
    wrapper = textwrap.TextWrapper(width=max_width)
    word_list = wrapper.wrap(text=string)
    result = ''
    for word in word_list:
        result = result + word + '\n'
    return result

if __name__ == '__main__':
    string, max_width = input(), int(input())
    result = wrap(string, max_width)
    print(result)

ABCDEFGHIJKLIMNOQRSTUVWXYZ
4
ABCD
EFGH
IJKL
IMNO
QRST
UVWX
YZ



## Designer Door Mat

In [41]:
if __name__ == '__main__':
    N, M = map(int,input().split())
    for i in range(1,N,2): 
        print((i * ".|.").center(M, "-"))
    print("WELCOME".center(M,"-"))
    for i in range(N-2,-1,-2): 
        print((i * ".|.").center(M, "-"))

9 27
------------.|.------------
---------.|..|..|.---------
------.|..|..|..|..|.------
---.|..|..|..|..|..|..|.---
----------WELCOME----------
---.|..|..|..|..|..|..|.---
------.|..|..|..|..|.------
---------.|..|..|.---------
------------.|.------------


## String Formatting

In [49]:
len(bin(17)[bin(17).find('b')+1:])

5

In [61]:
def print_formatted(number):
    # your code goes here
    pad = len(bin(number)[bin(number).find('b')+1:])
    space = ' '
    for i in range(number):
        num = i+1
        print(str(num).rjust(pad)+space+
              str(oct(num)[oct(num).find('o')+1:]).rjust(pad)+space
             + str(hex(num)[hex(num).find('x')+1:]).upper().rjust(pad)+space + 
              str(bin(num)[bin(num).find('b')+1:]).rjust(pad))

if __name__ == '__main__':
    n = int(input())
    print_formatted(n)

17
    1     1     1     1
    2     2     2    10
    3     3     3    11
    4     4     4   100
    5     5     5   101
    6     6     6   110
    7     7     7   111
    8    10     8  1000
    9    11     9  1001
   10    12     A  1010
   11    13     B  1011
   12    14     C  1100
   13    15     D  1101
   14    16     E  1110
   15    17     F  1111
   16    20    10 10000
   17    21    11 10001


In [59]:
def main():
    N = int(input())
    
    width = len(bin(N)[2:])
    for i in range(1, N + 1):
        print(' '.join(map(lambda x: x.rjust(width), [str(i), oct(i)[2:], hex(i)[2:].upper(), bin(i)[2:]])))

if __name__ == '__main__':
    main()

17
    1     1     1     1
    2     2     2    10
    3     3     3    11
    4     4     4   100
    5     5     5   101
    6     6     6   110
    7     7     7   111
    8    10     8  1000
    9    11     9  1001
   10    12     A  1010
   11    13     B  1011
   12    14     C  1100
   13    15     D  1101
   14    16     E  1110
   15    17     F  1111
   16    20    10 10000
   17    21    11 10001


## Alphabet Rangoli

In [63]:
import string
alpha = string.ascii_lowercase

n = int(input())
L = []
for i in range(n):
    s = "-".join(alpha[i:n])
    L.append((s[::-1]+s[1:]).center(4*n-3, "-"))
print('\n'.join(L[:0:-1]+L))

5
--------e--------
------e-d-e------
----e-d-c-d-e----
--e-d-c-b-c-d-e--
e-d-c-b-a-b-c-d-e
--e-d-c-b-c-d-e--
----e-d-c-d-e----
------e-d-e------
--------e--------


In [82]:
def print_rangoli(size):
    # your code goes here
    import string
    alpha = string.ascii_lowercase
    L = []
    for i in range(size):
        s = "-".join(alpha[i:size])
        L.append((s[::-1]+s[1:]).center(4*size-3, "-"))
    print('\n'.join(L[:0:-1]+L))

if __name__ == '__main__':
    n = int(input())
    print_rangoli(n)

5
--------e--------
------e-d-e------
----e-d-c-d-e----
--e-d-c-b-c-d-e--
e-d-c-b-a-b-c-d-e
--e-d-c-b-c-d-e--
----e-d-c-d-e----
------e-d-e------
--------e--------


## Capitalize!

In [84]:
'a'.isalpha()

True

In [85]:
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the solve function below.
def solve(s):
    name_old = s.split()
    name_new = []
    for name in name_old:
        if name[0].isalpha():
            name_new.append(name[0].upper()+name[1:])
#             name_new = name_new + name[0].upper()+name[1:]
        else:
            name_new.append(name)
    str1 = " "
    return str1.join(name_new)

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    s = input()

    result = solve(s)

    fptr.write(result + '\n')

    fptr.close()

KeyError: 'OUTPUT_PATH'

In [86]:
s = input()
for x in s[:].split():
    s = s.replace(x, x.capitalize())
print(s)

chris alan
Chris Alan


In [87]:
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the solve function below.
def solve(s):
    for x in s[:].split():
        s = s.replace(x, x.capitalize())
    return s
    
    
if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    s = input()

    result = solve(s)

    fptr.write(result + '\n')

    fptr.close()

KeyError: 'OUTPUT_PATH'

## The Minion Game

In [90]:
s = input()

vowels = 'AEIOU'

kevsc = 0
stusc = 0
for i in range(len(s)):
    if s[i] in vowels:
        kevsc += (len(s)-i)
    else:
        stusc += (len(s)-i)

if kevsc > stusc:
    print ("Kevin", kevsc)
elif kevsc < stusc:
    print ("Stuart", stusc)
else:
    print ("Draw")

BANANA
Stuart 12


In [91]:
def minion_game(string):
    # your code goes here
    vowels = 'AEIOU'

    kevsc = 0
    stusc = 0
    for i in range(len(s)):
        if s[i] in vowels:
            kevsc += (len(s)-i)
        else:
            stusc += (len(s)-i)

    if kevsc > stusc:
        print ("Kevin", kevsc)
    elif kevsc < stusc:
        print ("Stuart", stusc)
    else:
        print ("Draw")

if __name__ == '__main__':
    s = input()
    minion_game(s)

BANANA
Stuart 12


## Merge the Tools!

In [98]:
def merge_the_tools(string, k):
    # your code goes here
    group = len(string)//k
    for i in range(group):
        substring = string[i*k:(i+1)*k]
#         print(substring)
        result = []
        for char in substring:
            if substring.count(char)>1:
                if char not in result:
                    result.append(char)
            else:
                result.append(char)
        result_str = ""
        print(result_str.join(result))

if __name__ == '__main__':
    string, k = input(), int(input())
    merge_the_tools(string, k)

AABCAAADA
3
AB
CA
AD


# Sets

## Introduction to Sets

In [104]:
def average(array):
    # your code goes here
    cal_set = set(array)
#     print(cal_set)
    return round(sum(cal_set)/len(cal_set),3)
    
if __name__ == '__main__':
    n = int(input())
    arr = list(map(int, input().split()))
    result = average(arr)
    print(result)

10
161 182 161 154 176 170 167 171 170 174
169.375


## Symmetric Difference

In [66]:
if __name__ == '__main__':
    M = int(input())
    a = set(list(map(int,input().split())))
    N = int(input())
    b = set(list(map(int,input().split())))
    diff_a = list(a.difference(b))
    diff_b = list(b.difference(a))
    diff = sorted(diff_a + diff_b)
    for i in diff:
        print(i)

4
2 4 5 9
4
2 4 11 12
5
9
11
12


In [ ]:
a,b=(int(input()),input().split())
c,d=(int(input()),input().split())
x=set(b)
y=set(d)
p=y.difference(x)
q=x.difference(y)
r=p.union(q)
print ('\n'.join(sorted(r, key=int)))

## No Idea!

In [ ]:
if __name__ == '__main__':
    M = int(input())
    a = set(list(map(int,input().split())))
    N = int(input())
    b = set(list(map(int,input().split())))

In [69]:
if __name__ == '__main__':
    n, m = list(map(int,input().split()))
    arr = list(map(int,input().split()))
    A = list(map(int,input().split()))
    B = list(map(int,input().split()))
    happy = 0
    for i in arr:
        if i in A:
            happy += 1
        if i in B:
            happy -= 1
    print(happy)

3 2
1 5 3
3 1
5 7
1


In [72]:
n, m = input().split()

sc_ar = input().split()

A = set(input().split())
B = set(input().split())
print (sum([(i in A) - (i in B) for i in sc_ar]))

3 2
1 5 3
3 1
5 7
1


## Set .add()



In [74]:
if __name__ == '__main__':
    n = int(input())
    stamp = set()
    for i in range(n):
        stamp.add(input())
    print(len(stamp))

7
UK
CH
USA
FR
NZ
UK
FR
5


## Set .discard(), .remove() & .pop()

In [78]:
if __name__ == '__main__':
    n = int(input())
    arr = set((map(int,input().split())))
    N = int(input())
    for i in range(N):
        command, *num = input().split()
        if command == 'pop':
            arr.pop()
        if command == 'remove':
            arr.remove(num)
        if command == 'discard':
            arr.discard(num)    

9
1 2 3 4 5 6 7 8 9
10
pop
remove 9


TypeError: unhashable type: 'list'

In [80]:
n = int(input())
s = set(map(int, input().split())) 
for i in range(int(input())):
    eval('s.{0}({1})'.format(*input().split()+['']))

print(sum(s))

9
1 2 3 4 5 6 7 8 9
10
pop
remove 9
discard 9
discard 8
remove 7
pop
discard 6
remove 5
pop
discard 5
4


In [83]:
n = int(input())
s = set(map(int, input().split()))
m = int(input())
for  i in range(m):
    attr, *kw = input().split()
    if kw:
        getattr(s,attr)(*(map(int, *kw)))
    else:
        getattr(s,attr)()
print(sum(s))

9
1 2 3 4 5 6 7 8 9
10
pop
remove 9
discard 9
discard 8
remove 7
pop
discard 6
remove 5
pop
discard 5
4


## Set .union() Operation

In [84]:
if __name__ == '__main__':
    n = int(input())
    a = set(map(int, input().split()))
    m = int(input())
    b = set(map(int, input().split()))
    print(len(a.union(b)))

9
1 2 3 4 5 6 7 8 9
9
10 1 2 3 11 21 55 6 8
13


## Set .intersection() Operation

In [86]:
if __name__ == '__main__':
    n = int(input())
    a = set(map(int, input().split()))
    m = int(input())
    b = set(map(int, input().split()))
    print(len(a.intersection(b)))

9
1 2 3 4 5 6 7 8 9
9
10 1 2 3 11 21 55 6 8
5


## Set .difference() Operation

In [87]:
if __name__ == '__main__':
    n = int(input())
    a = set(map(int, input().split()))
    m = int(input())
    b = set(map(int, input().split()))
    print(len(a.difference(b)))

9
1 2 3 4 5 6 7 8 9
9
10 1 2 3 11 21 55 6 8
4


## Set Mutations

In [1]:
if __name__ == '__main__':
    (_, A) = (int(input()),set(map(int, input().split())))
    B = int(input())
    for _ in range(B):
        (command, newSet) = (input().split()[0],set(map(int, input().split())))
        getattr(A, command)(newSet)

    print (sum(A))

16
1 2 3 4 5 6 7 8 9 10 11 12 13 14 24 52
4
intersection_update 10
2 3 5 6 8 9 1 4 7 11
update 2
55 66
symmetric_difference_update 5
22 7 35 62 58
difference_update 7
11 22 35 55 58 62 66
38


In [2]:
_

3

In [3]:
A

{1, 2, 3, 4, 5, 6, 8, 9}

In [4]:
command

'difference_update'

In [5]:
newSet

{11, 22, 35, 55, 58, 62, 66}

## The Captain's Room



In [15]:
if __name__ == '__main__':
    (k, room) = (int(input()), list(map(int, input().split())))
    room_set = set(room)
    for i in room_set:
        if room.count(i) == 1:
            print(i)

5
1 2 3 6 5 4 4 2 5 3 6 1 6 5 3 2 4 1 2 5 1 4 3 6 8 4 3 1 5 6 2
8


In [17]:
if __name__ == '__main__':
    k,arr = int(input()),list(map(int, input().split()))

    myset = set(arr)

    print(((sum(myset)*k)-(sum(arr)))//(k-1))

5
1 2 3 6 5 4 4 2 5 3 6 1 6 5 3 2 4 1 2 5 1 4 3 6 8 4 3 1 5 6 2
8


In [18]:
sum(myset)*k

145

In [19]:
sum(arr)

113

In [20]:
k-1

4

In [22]:
((sum(myset)*k)-(sum(arr)))//(k-1)

8

In [30]:
arr = [1,1,1,2,2,2,3,4,4,4]
arr

[1, 1, 1, 2, 2, 2, 3, 4, 4, 4]

In [31]:
myset = set(arr)
myset

{1, 2, 3, 4}

In [32]:
k = 3
k

3

In [33]:
sum(myset)*k

30

In [34]:
sum(arr)

24

In [35]:
((sum(myset)*k)-(sum(arr)))//(k-1)

3

In [36]:
k, rooms, single, multiple = input(), input().split(), set(), set()

for room in rooms: single.add(room) if room not in single else multiple.add(room)

print(single.difference(multiple).pop())

5
1 2 3 6 5 4 4 2 5 3 6 1 6 5 3 2 4 1 2 5 1 4 3 6 8 4 3 1 5 6 2
8


## Check Subset

In [38]:
if __name__ == '__main__':
    T = int(input())
    for i in range(T):
        nA = int(input())
        A = set(map(int, input().split()))
        nB = int(input())
        B = set(map(int, input().split()))
        print(A.issubset(B))

3
5
1 2 3 5 6
9
9 8 5 6 3 2 1 4 7
True
1
2
5
3 6 5 4 1
False
7
1 2 3 5 6 8 9
3
9 8 2
False


In [ ]:
for _ in range(int(input())):
x, a, z, b = input(), set(input().split()), input(), set(input().split())
print(a.issubset(b))

## Check Strict Superset



In [42]:
def isstrictsuperset(a,b):
    # true if a is a strict superset of b
    return b.issubset(a) and not(a.issubset(b))

a = set(int(x) for x in input().split(' '))
n = int(input())
res = True

for _ in range(n):
    b = set(int(x) for x in input().split(' '))
    res &= isstrictsuperset(a,b)
    
print(res)
                                                  

1 2 3 4 5 6 7 8 9 10 11 12 23 45 84 78
2
1 2 3 4 5
100 11 12
False


In [54]:
if __name__ == '__main__':
    a, n =  set(map(int, input().split())), int(input())
    b = set()
    for i in range(n):
        b = b.union(set(map(int, input().split())))   
    print(a.issuperset(b))

1 2 3 4 5 6 7 8 9 10 11 12 23 45 84 78
2
1 2 3 4 5
100 11 12
False


# Numpy

## Arrays

In [66]:
if __name__ == '__main__':
    import numpy as np
    print(np.array(list(map(int, input().split(' '))), float)[::-1])

1 2 3 4 -8 -10
[-10.  -8.   4.   3.   2.   1.]


In [68]:
import numpy

def arrays(arr):
    # complete this function
    # use numpy.array
    result = numpy.array(arr, float)[::-1]
    return result

arr = input().strip().split(' ')
result = arrays(arr)
print(result)

1 2 3 4 -8 -10
[-10.  -8.   4.   3.   2.   1.]


## Shape and Reshape



In [73]:
import numpy

arr = list(map(int,input().strip().split(' ')))
arr = numpy.array(arr)
arr.shape = (3,3)
print(arr)

1 2 3 4 5 6 7 8 9
[[1 2 3]
 [4 5 6]
 [7 8 9]]


## Transpose and Flatten

In [76]:
import numpy

n, m = map(int, input().split(' '))
for i in range(m):
    arr = numpy.array(map(int,input().strip().split(' ')))
    
print(arr.transpose())
print(arr.flatten())

2 2
1 2


In [77]:
n, m = map(int, input().split())
array = numpy.array([input().strip().split() for _ in range(n)], int)
print (array.transpose())
print (array.flatten())

2 2
1 2
3 4
[[1 3]
 [2 4]]
[1 2 3 4]


## Concatenate



In [80]:
import numpy
n, m, p = map(int, input().split(' '))
a = numpy.array([input().strip().split() for _ in range(n)], int)
b = numpy.array([input().strip().split() for _ in range(m)], int)
print(numpy.concatenate((a,b), axis=0))

4 3 2
1 2
1 2
1 2
1 2
3 4
3 4
3 4
[[1 2]
 [1 2]
 [1 2]
 [1 2]
 [3 4]
 [3 4]
 [3 4]]


In [81]:
import numpy as np
a, b, c = map(int,input().split())
arrA = np.array([input().split() for _ in range(a)],int)
arrB = np.array([input().split() for _ in range(b)],int)
print(np.concatenate((arrA, arrB), axis = 0))

4 3 2
1 2
1 2
1 2
1 2
3 4
3 4
3 4
[[1 2]
 [1 2]
 [1 2]
 [1 2]
 [3 4]
 [3 4]
 [3 4]]


## Zeros and Ones

In [85]:
import numpy
nums = tuple(map(int, input().split()))
print (numpy.zeros(nums, dtype = numpy.int))
print (numpy.ones(nums, dtype = numpy.int))

3 3 3
[[[0 0 0]
  [0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]
  [0 0 0]]]
[[[1 1 1]
  [1 1 1]
  [1 1 1]]

 [[1 1 1]
  [1 1 1]
  [1 1 1]]

 [[1 1 1]
  [1 1 1]
  [1 1 1]]]


In [86]:
import numpy
nums = list(map(int, input().split()))
print (numpy.zeros(nums, dtype = numpy.int))
print (numpy.ones(nums, dtype = numpy.int))

3 3 3
[[[0 0 0]
  [0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]
  [0 0 0]]]
[[[1 1 1]
  [1 1 1]
  [1 1 1]]

 [[1 1 1]
  [1 1 1]
  [1 1 1]]

 [[1 1 1]
  [1 1 1]
  [1 1 1]]]


## Eye and Identity



In [89]:
import numpy

numpy.set_printoptions(sign=' ')
print(numpy.eye(*map(int, input().split())))

3 6
[[ 1.  0.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.  0.]]


In [91]:
import numpy
print(str(numpy.eye(*map(int,input().split()))).replace('1',' 1').replace('0',' 0'))

3 6
[[  1.   0.   0.   0.   0.   0.]
 [  0.   1.   0.   0.   0.   0.]
 [  0.   0.   1.   0.   0.   0.]]


## Array Mathematics

In [96]:
import numpy as np

n, m = map(int, input().split())
a, b = (np.array([input().split() for _ in range(n)], dtype=int) for _ in range(2))
print(a+b, a-b, a*b, a//b, a%b, a**b, sep='\n')

1 4
1 2 3 4
5 6 7 8
[[ 6  8 10 12]]
[[-4 -4 -4 -4]]
[[ 5 12 21 32]]
[[0 0 0 0]]
[[1 2 3 4]]
[[    1    64  2187 65536]]


## Floor, Ceil and Rint

In [98]:
import numpy as np
np.set_printoptions(legacy='1.13')
arr = np.array(input().split(), float)
print(np.floor(arr))
print(np.ceil(arr))
print(np.rint(arr))

1.1 2.2 3.3 4.4 5.5 6.6 7.7 8.8 9.9
[ 1.  2.  3.  4.  5.  6.  7.  8.  9.]
[  2.   3.   4.   5.   6.   7.   8.   9.  10.]
[  1.   2.   3.   4.   6.   7.   8.   9.  10.]


In [100]:
import numpy as np
np.set_printoptions(legacy='1.13')
arr = list(map(float,input().split()))
arr = np.array(arr)
print(np.floor(arr))
print(np.ceil(arr))
print(np.rint(arr))

1.1 2.2 3.3 4.4 5.5 6.6 7.7 8.8 9.9
[ 1.  2.  3.  4.  5.  6.  7.  8.  9.]
[  2.   3.   4.   5.   6.   7.   8.   9.  10.]
[  1.   2.   3.   4.   6.   7.   8.   9.  10.]


## Sum and Prod

In [102]:
import numpy as np

n, m = map(int, input().split())
arr = (np.array([input().split() for _ in range(n)], dtype=int))
print(np.prod(np.sum(arr, axis=0)))

2 2
1 2
3 4
24


## Min and Max

In [106]:
import numpy as np

n, m = map(int, input().split())
arr = (np.array([input().split() for _ in range(n)], dtype=int))
print(np.max(np.min(arr, axis=1)))

4 2
2 5
3 7
1 3
4 0
3


## Mean, Var, and Std

In [110]:
import numpy as np
# np.set_printoptions(legacy='1.13')
n, m = map(int, input().split())
arr = (np.array([input().split() for _ in range(n)], dtype=int))
arr = np.around(arr,decimals=11)
print(np.mean(arr, axis=1))
print(np.var(arr, axis=0))
print(np.std(arr))

2 2
1 2
3 4
[ 1.5  3.5]
[ 1.  1.]
1.11803398875


In [111]:
import numpy as np
# np.set_printoptions(legacy='1.13')
n, m = map(int, input().split())
arr = (np.array([input().split() for _ in range(n)], dtype=int))
print(np.mean(arr, axis=1))
print(np.var(arr, axis=0))
print(np.around(np.std(arr),decimals=11))

2 2
1 2
3 4
[ 1.5  3.5]
[ 1.  1.]
1.11803398875


## Dot and Cross


In [119]:
import numpy as np

n = int(input())
a = (np.array([input().split() for _ in range(n)], dtype=int))
b = (np.array([input().split() for _ in range(n)], dtype=int))
print (np.dot(a, b))

2
1 2
3 4
1 2
3 4
[[ 7 10]
 [15 22]]


## Inner and Outer

In [123]:
import numpy as np
A = (np.array(input().split(), int))
B = (np.array(input().split(), int))
print(np.inner(A, B))
print(np.outer(A, B))

0 1
2 3
3
[[0 0]
 [2 3]]


## Polynomials

In [125]:
import numpy as np
P = (np.array(input().split(), float))
x = int(input())
print(np.polyval(P, x))

1.1 2 3
0
3.0


## Linear Algebra

In [130]:
import numpy as np

n = int(input())
a = (np.array([input().split() for _ in range(n)], dtype=float))
print(round(np.linalg.det(a), 2))

2
1.1 1.1
1.1 1.1
0.0


In [128]:
import numpy
print(round(numpy.linalg.det(numpy.array([list(map(float,input().split())) for _ in range(int(input()))])),2))

2
1.1 1.1
1.1 1.1
0.0


# Date & Time

## Calendar Module

In [135]:
import calendar

print (calendar.TextCalendar(firstweekday=6).formatyear(2021))

                                  2021

      January                   February                   March
Su Mo Tu We Th Fr Sa      Su Mo Tu We Th Fr Sa      Su Mo Tu We Th Fr Sa
                1  2          1  2  3  4  5  6          1  2  3  4  5  6
 3  4  5  6  7  8  9       7  8  9 10 11 12 13       7  8  9 10 11 12 13
10 11 12 13 14 15 16      14 15 16 17 18 19 20      14 15 16 17 18 19 20
17 18 19 20 21 22 23      21 22 23 24 25 26 27      21 22 23 24 25 26 27
24 25 26 27 28 29 30      28                        28 29 30 31
31

       April                      May                       June
Su Mo Tu We Th Fr Sa      Su Mo Tu We Th Fr Sa      Su Mo Tu We Th Fr Sa
             1  2  3                         1             1  2  3  4  5
 4  5  6  7  8  9 10       2  3  4  5  6  7  8       6  7  8  9 10 11 12
11 12 13 14 15 16 17       9 10 11 12 13 14 15      13 14 15 16 17 18 19
18 19 20 21 22 23 24      16 17 18 19 20 21 22      20 21 22 23 24 25 26
25 26 27 28 29 30         23 24 

In [136]:
import calendar

n1,n2,n3=map(int,input().split())
print((calendar.day_name[calendar.weekday(n3,n1,n2)]).upper())

08 05 2015
WEDNESDAY


## Time Delta

In [137]:
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the time_delta function below.
def time_delta(t1, t2):
    
    return str(int(abs((t1-t2).total_seconds())))
    
if __name__ == '__main__':
    from datetime import datetime
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    t = int(input())

    for t_itr in range(t):
        t1 = datetime.strptime(input(),'%a %d %b %Y %H:%M:%S %z')

        t2 = datetime.strptime(input(),'%a %d %b %Y %H:%M:%S %z')

        delta = time_delta(t1, t2)

        fptr.write(delta + '\n')

    fptr.close()


KeyError: 'OUTPUT_PATH'

# Regex and Parsing

## Re.findall() & Re.finditer()

In [138]:
import re
v = "aeiou"
c = "qwrtypsdfghjklzxcvbnm"
m = re.findall(r"(?<=[%s])([%s]{2,})[%s]" % (c, v, c), input(), flags = re.I)
print('\n'.join(m or ['-1']))

rabcdeefgyYhFjkIoomnpOeorteeeeet
ee
Ioo
Oeo
eeeee


## Re.start() & Re.end()

In [141]:
import re

S = input()
k = input()

pattern = re.compile(k)
r = pattern.search(S)
if not r: print ("(-1, -1)")
while r:
    print ("({0}, {1})".format(r.start(), r.end() - 1))
    r = pattern.search(S,r.start() + 1)

aaadaa
aa
(0, 1)
(1, 2)
(4, 5)


## Regex Substitution

In [144]:
import re

N = int(input())

for i in range(N):
    print (re.sub(r'(?<= )(&&|\|\|)(?= )', lambda x: 'and' if x.group() == '&&' else 'or', input()))

11
a = 1;
a = 1;
b = input();
b = input();
if a + b > 0 && a - b < 0:
if a + b > 0 and a - b < 0:
    start()
    start()
elif a*b > 10 || a/b < 1:
elif a*b > 10 or a/b < 1:
    stop()
    stop()
print set(list(a)) | set(list(b)) 
print set(list(a)) | set(list(b)) 
#Note do not change &&& or ||| or & or |
#Note do not change &&& or ||| or & or |
#Only change those '&&' which have space on both sides.
#Only change those '&&' which have space on both sides.
#Only change those '&&' which have space on both sides.
#Only change those '&&' which have space on both sides.




## Validating Roman Numerals

In [145]:
thousand = "(?:(M){0,3})?"
hundred  = "(?:(D?(C){0,3})|(CM)|(CD))?"
ten      = "(?:(L?(X){0,3})|(XC)|(XL))?"
unit     = "(?:(V?(I){0,3})|(IX)|(IV))?"

regex_pattern = r"^" + thousand + hundred + ten + unit + "$"	# Do not delete 'r'.

import re
print(str(bool(re.match(regex_pattern, input()))))

CDXXI
True


In [146]:
thousand = 'M{0,3}'
hundred = '(C[MD]|D?C{0,3})'
ten = '(X[CL]|L?X{0,3})'
digit = '(I[VX]|V?I{0,3})'
regex_pattern = r"%s%s%s%s$" % (thousand, hundred, ten, digit)   # Do not delete 'r'

import re
print(str(bool(re.match(regex_pattern, input()))))

CDXXI
True


## Validating phone numbers

In [147]:
import re

n = int(input())
pattern = re.compile(r'^[7-9]\d{9}$')
numbers = [input() for _ in range(n)]

for number in numbers:
    if pattern.match(number):
        print('YES')
    else:
        print('NO')

2
9587456281
1252478965
YES
NO


In [148]:
numbers

['9587456281', '1252478965']

In [149]:
number

'1252478965'

## Validating and Parsing Email Addresses



In [150]:
import re
n = int(input())
for _ in range(n):
    x, y = input().split(' ')
    m = re.match(r'<[A-Za-z](\w|-|\.|_)+@[A-Za-z]+\.[A-Za-z]{1,3}>', y)
    if m:
        print(x,y)

2
DEXTER <dexter@hotmail.com>
DEXTER <dexter@hotmail.com>
VIRUS <virus!@variable.:p>


## Hex Color Code


In [151]:
import re

N = int(input())
for i in range(N):
    s=input()

    x=s.split()

    if len(x)>1  and  '{' not in x:
        x=re.findall(r'#[a-fA-F0-9]{3,6}',s)
        [print(i) for  i in x]
# (?<!^)(#(?:[\da-f]{3}){1,2})

11
#BED
{
    color: #FfFdF8; background-color:#aef;
#FfFdF8
#aef
    font-size: 123px;
    background: -webkit-linear-gradient(top, #f9f9f9, #fff);
#f9f9f9
#fff
}
#Cab
{
    background-color: #ABC;
#ABC
    border: 2px dashed #fff;
#fff
}


## HTML Parser - Part 1



In [153]:
from html.parser import HTMLParser

class MyHTMLParser(HTMLParser):
    def handle_starttag(self, tag, attrs):        
        print ('Start :',tag)
        for ele in attrs:
            print ('->',ele[0],'>',ele[1])
            
    def handle_endtag(self, tag):
        print ('End   :',tag)
        
    def handle_startendtag(self, tag, attrs):
        print ('Empty :',tag)
        for ele in attrs:
            print ('->',ele[0],'>',ele[1])
            
MyParser = MyHTMLParser()
MyParser.feed(''.join([input().strip() for _ in range(int(input()))]))

2
<html><head><title>HTML Parser - I</title></head>
<body data-modal-target class='1'><h1>HackerRank</h1><br /></body></html>
Start : html
Start : head
Start : title
End   : title
End   : head
Start : body
-> data-modal-target > None
-> class > 1
Start : h1
End   : h1
Empty : br
End   : body
End   : html


## HTML Parser - Part 2



In [156]:
from html.parser import HTMLParser

class MyHTMLParser(HTMLParser):
    def handle_comment(self, data):
        number_of_line = len(data.split('\n'))
        if number_of_line>1:
            print('>>> Multi-line Comment')
        else:
            print('>>> Single-line Comment')
        if data.strip():
            print(data)

    def handle_data(self, data):
        if data.strip():
            print(">>> Data")
            print(data)
  
html = ""       
for i in range(int(input())):
    html += input().rstrip()
    html += '\n'
    
parser = MyHTMLParser()
parser.feed(html)
parser.close()

4
<!--[if IE 9]>IE9-specific content
<![endif]-->
<div> Welcome to HackerRank</div>
<!--[if IE 9]>IE9-specific content<![endif]-->
>>> Multi-line Comment
[if IE 9]>IE9-specific content
<![endif]
>>> Data
 Welcome to HackerRank
>>> Single-line Comment
[if IE 9]>IE9-specific content<![endif]


## Detect HTML Tags, Attributes and Attribute Values

In [157]:
from html.parser import HTMLParser
class MyHTMLParser(HTMLParser):
    def handle_starttag(self, tag, attrs):
        print(tag)
        [print('-> {} > {}'.format(*attr)) for attr in attrs]
        
html = '\n'.join([input() for _ in range(int(input()))])
parser = MyHTMLParser()
parser.feed(html)
parser.close()

9
<head>
<title>HTML</title>
</head>
<object type="application/x-flash" 
  data="your-file.swf" 
  width="0" height="0">
  <!-- <param name="movie" value="your-file.swf" /> -->
  <param name="quality" value="high"/>
</object>
head
title
object
-> type > application/x-flash
-> data > your-file.swf
-> width > 0
-> height > 0
param
-> name > quality
-> value > high


## Validating UID

In [158]:
import re

for _ in range(int(input())):
    u = ''.join(sorted(input()))
    try:
        assert re.search(r'[A-Z]{2}', u)
        assert re.search(r'\d\d\d', u)
        assert not re.search(r'[^a-zA-Z0-9]', u)
        assert not re.search(r'(.)\1', u)
        assert len(u) == 10
    except:
        print('Invalid')
    else:
        print('Valid')

2
B1CD102354
Invalid
B1CDEF2354
Valid


In [159]:
u

'12345BCDEF'

## Validating Credit Card Numbers



In [160]:
import re
TESTER = re.compile(
    r"^"
    r"(?!.*(\d)(-?\1){3})"
    r"[456]"
    r"\d{3}"
    r"(?:-?\d{4}){3}"
    r"$")
for _ in range(int(input().strip())):
    print("Valid" if TESTER.search(input().strip()) else "Invalid")

6
4123456789123456
Valid
5123-4567-8912-3456
Valid
61234-567-8912-3456
Invalid
4123356789123456
Valid
5133-3367-8912-3456
Invalid
5123 - 3567 - 8912 - 3456
Invalid


In [161]:
import re

def check(card):
    if not re.search("^[456]\d{3}(-?\d{4}){3}$",card) or re.search(r"(\d)\1{3}", re.sub("-", "",card)):
        return False
    return True

for i in range(int(input())):
    print("Valid" if check(input()) else "Invalid")

6
4123456789123456
Valid
5123-4567-8912-3456
Valid
61234-567-8912-3456
Invalid
4123356789123456
Valid
5133-3367-8912-3456
Invalid
5123 - 3567 - 8912 - 3456
Invalid


## Validating Postal Codes

In [162]:
import re
s=input()
print (bool(re.match(r'^[1-9][\d]{5}$',s) and len(re.findall(r'(\d)(?=\d\1)',s))<2 ))

110000
False


In [163]:
regex_integer_in_range = r"^[1-9][0-9]{5}$"
regex_alternating_repetitive_digit_pair = r"(\d)(?=.\1)"

import re
P = input()

print (bool(re.match(regex_integer_in_range, P)) 
and len(re.findall(regex_alternating_repetitive_digit_pair, P)) < 2)

110000
False


## Group(), Groups() & Groupdict()



In [164]:
import re
m = re.search(r'([a-zA-Z0-9])\1+', input().strip())
print(m.group(1) if m else -1)

..12345678910111213141516171820212223
1


## Matrix Script



In [166]:
#!/bin/python3

import math
import os
import random
import re
import sys




first_multiple_input = input().rstrip().split()

n = int(first_multiple_input[0])

m = int(first_multiple_input[1])

matrix = []

for _ in range(n):
    matrix_item = input()
    matrix.append(matrix_item)

encoded_string = "".join([matrix[j][i] for i in range(m) for j in range(n)])
pat = r'(?<=[a-zA-Z0-9])[^a-zA-Z0-9]+(?=[a-zA-Z0-9])'
print(re.sub(pat,' ',encoded_string))

7 3
Tsi
h%x
i #
sM 
$a 
#t%
ir!
This is Matrix#  %!


In [167]:
matrix

['Tsi', 'h%x', 'i #', 'sM ', '$a ', '#t%', 'ir!']

In [168]:
encoded_string

'This$#is% Matrix#  %!'

# Math

## Polar Coordinates

In [11]:
import cmath

num = complex(input())
print(abs(num))
print(cmath.phase(num))

1+2j
2.23606797749979
1.1071487177940904


In [4]:
import cmath
print(*cmath.polar(complex(input())), sep='\n')

1+2j
2.23606797749979
1.1071487177940904


## Find Angle MBC

In [19]:
import math

ab = int(input())
bc = int(input())

print(str(int(round(math.degrees(math.atan2(ab,bc)))))+chr(176))

10
10
45°


## Triangle Quest 2

Demlo numbers
The output requires the sequence of Demlo numbers. For n<=9, the squares of the first few repunits are precisely the Demlo numbers. Example: 1^2=1, 11^2=121, 111^2=12321 and so on.

At first lets see how to get repunits. (10^i - 1) / 9 results in the repunits sequence: 1, 11, 111 etc.

Then calculating square of each repunits, we can get Demlo numbers.

So, the code to perform these operations to generate Demlo numbers is:

In [27]:
for i in range(1,int(input())+1): #More than 2 lines will result in 0 score. Do not leave a blank line also
    print (int(((10**i-1)//9)**2))

9
1
121
12321
1234321
123454321
12345654321
1234567654321
123456787654321
12345678987654321


## Mod Divmod 

In [29]:
a = int(input())
b = int(input())
print(a//b)
print(a%b)
print(divmod(a,b))

177
10
17
7
(17, 7)


## Power - Mod Power



In [30]:
a = int(input())
b = int(input())
m = int(input())
print(pow(a,b))
print(pow(a,b,m))

3
4
5
81
1


## Integers Come In All Sizes



In [31]:
a = int(input())
b = int(input())
c = int(input())
d = int(input())
print(a**b + c**d)

9
29
7
27
4710194409608608369201743232


## Triangle Quest



In [34]:
for i in range(1,int(input())):
    print((10**(i)//9)*i)

5
1
22
333
4444


# Itertools

## itertools.product()

In [12]:
import itertools
a = list(map(int,input().split()))
b = list(map(int,input().split()))
print (*itertools.product(a,b))

1 2
3 4
(1, 3) (1, 4) (2, 3) (2, 4)


## itertools.permutations()



In [16]:
import itertools
string, n = input().split()
print(*[''.join(i) for i in itertools.permutations(sorted(string),int(n))],sep='\n')

HACK 2
AC
AH
AK
CA
CH
CK
HA
HC
HK
KA
KC
KH


## itertools.combinations()



In [20]:
import itertools
string, n = input().split()
string = sorted(string)
for i in range(1, int(n)+1):
    for j in itertools.combinations(string, i):
        print (''.join(j))

HACK 2
A
C
H
K
AC
AH
AK
CH
CK
HK


## itertools.combinations_with_replacement()



In [21]:
import itertools
string, n = input().split()
print(*[''.join(i) for i in itertools.combinations_with_replacement(sorted(string),int(n))],sep='\n')

HACK 2
AA
AC
AH
AK
CC
CH
CK
HH
HK
KK


## Compress the String!



Sure thing. Basically I'm unpacking the list comprehension and printing each element of it.

The list is built of elements of (len(list(c)), int(k)). len(list(c)) is simply the number of occurences of a character c returned by the groupby function. k is just the key value, the character itself.

In [24]:
from itertools import groupby
print(*[(len(list(c)), int(k)) for k, c in groupby(input())])

1222311
(1, 1) (3, 2) (1, 3) (2, 1)


In [ ]:
# [k for k, g in groupby('AAAABBBCCDAABBB')] --> A B C D A B
# [list(g) for k, g in groupby('AAAABBBCCD')] --> AAAA BBB CC D

## Iterables and Iterators



In [25]:
from itertools import combinations

_,s,n = input(),input().split(),int(input())
t = list(combinations(s,n))
f = [i for i in t if 'a' in i]
print(len(f)/len(t))

4
a a c d
2
0.8333333333333334


## Maximize It!



In [30]:
n, m = map(int, input().split())
result = 0
for _ in range(n):
    num = list(map(int, input().split()))[1:]
    max_num = max(num)
    result = result + max_num**2
print(result%m)

3 1000
2 5 4
3 7 8 9
5 5 7 8 9 10
206


In [29]:
from itertools import product

K,M = map(int,input().split())
N = (list(map(int, input().split()))[1:] for _ in range(K))
results = map(lambda x: sum(i**2 for i in x)%M, product(*N))
print(max(results))

3 1000
2 5 4
3 7 8 9
5 5 7 8 9 10
206


In [31]:
from itertools import product
K,M = map(int,input().split())
nums = []
for _ in range(K):
    row = map(int,input().split()[1:])
    nums.append(map(lambda x:x**2%M, row))
print(max(map(lambda x: sum(x)%M, product(*nums))))

3 1000
2 5 4
3 7 8 9
5 5 7 8 9 10
206


# Collections

## collections.Counter()



In [2]:
from collections import Counter
n = int(input())
shoes = Counter(list(map(int, input().split())))
cus = int(input())
income = 0
for _ in range(cus):
    buy_shoe = list(map(int, input().split()))
#     print(buy_shoe)
    if shoes[buy_shoe[0]]:
#         print(shoes[buy_shoe[0]])
        shoes[buy_shoe[0]] -= 1
#         print(shoes[buy_shoe[0]])
#         print(income)
        income += buy_shoe[1]
#         print(income)
print(income)

10
2 3 4 5 6 8 7 6 5 18
6
6 55
6 45
6 55
4 40
18 60
10 50
200


In [3]:
shoes

Counter({2: 1, 3: 1, 4: 0, 5: 2, 6: 0, 8: 1, 7: 1, 18: 0})

In [4]:
cus

6

In [5]:
buy_shoe

[10, 50]

In [6]:
buy_shoe[0]

10

In [7]:
shoes[buy_shoe[0]]

0

In [8]:
shoes[buy_shoe[1]]

0

In [9]:
income

200

## DefaultDict Tutorial



In [20]:
from collections import defaultdict
n, m = map(int,input().split())
d = defaultdict(list)

for _ in range(n):
    d['A'].append(input())
for _ in range(m):
    d['B'].append(input())



5 2
a
a
b
a
b
a
b


In [21]:
d

defaultdict(list, {'A': ['a', 'a', 'b', 'a', 'b'], 'B': ['a', 'b']})

In [15]:
d['A']

['a', 'a', 'b', 'a', 'b']

In [16]:
d['B']

['a', 'b']

In [13]:
d.items()

dict_items([('a', ['a', 'a', 'b', 'a', 'b']), ('b', ['a', 'b'])])

In [ ]:
d

In [23]:
from collections import defaultdict
d = defaultdict(list)

n, m = map(int, input().split())

for i in range(1,n+1):
    d[input()].append(str(i))
    
for i in range(m):
    print (' '.join(d[input()]) or -1)

5 2
a
a
b
a
b
a
1 2 4
b
3 5


In [24]:
d

defaultdict(list, {'a': ['1', '2', '4'], 'b': ['3', '5']})

In [25]:
d['a']

['1', '2', '4']

In [26]:
d['b']

['3', '5']

## Collections.namedtuple()



In [ ]:
from collections import namedtuple
n, cols = int(input()), input()
Students = namedtuple('Students', cols, Class = 'Y')
for _ in range(n):
    id_no, marks, name, class_no = input.split()
    
    
    

In [27]:
from collections import namedtuple
n, Score = int(input()) , namedtuple('Score',input().split())
scores = [Score(*input().split()).MARKS for i in range(n)]
print("%.2f"% (sum(map(int,scores))/n))

5
ID MARKS NAME CLASS
1 97 Raymond 7
2 50 Steven 4
3 91 Adrian 9
4 72 Stewart 5
5 80 Peter 6
78.00


In [28]:
scores

['97', '50', '91', '72', '80']

In [29]:
Score

__main__.Score

In [30]:
n

5

In [33]:
Score.MARKS

In [35]:
sum(map(int,scores))

390

## Collections.OrderedDict()

In [ ]:
from collections import OrderedDict

supermarket_items = OrderedDict()
for _ in range(int(input())):
    supermarket_items[input()]


In [37]:
from collections import OrderedDict
d = OrderedDict()
for _ in range(int(input())):
    item, space, quantity = input().rpartition(' ')
    d[item] = d.get(item, 0) + int(quantity)
for item, quantity in d.items():
    print(item, quantity)

9
BANANA FRIES 12
POTATO CHIPS 30
APPLE JUICE 10
CANDY 5
APPLE JUICE 10
CANDY 5
CANDY 5
CANDY 5
POTATO CHIPS 30
BANANA FRIES 12
POTATO CHIPS 60
APPLE JUICE 20
CANDY 20


In [40]:
d

OrderedDict([('BANANA FRIES', 12),
             ('POTATO CHIPS', 60),
             ('APPLE JUICE', 20),
             ('CANDY', 20)])

In [41]:
item

'CANDY'

In [42]:
d.get(item, 0)

20

## Word Order

In [57]:
from collections import OrderedDict
d = OrderedDict()
for _ in range(int(input())):
    word = input()
    d[word] = d.get(word, 0) + 1
print(len(d.keys()))
print(*d.values())

4
bcdef
abcdefg
bcde
bcdef
3
2 1 1


In [44]:
d

OrderedDict([('bcdef', 2), ('abcdefg', 1), ('bcde', 1)])

In [48]:
len(d.keys())

3

In [49]:
d.items()

odict_items([('bcdef', 2), ('abcdefg', 1), ('bcde', 1)])

In [55]:
' '.join(str(d.values()))

'o d i c t _ v a l u e s ( [ 2 ,   1 ,   1 ] )'

In [56]:
print(*d.values())

2 1 1


In [59]:
from collections import Counter, OrderedDict
class OrderedCounter(Counter, OrderedDict):
    pass
d = OrderedCounter(input() for _ in range(int(input())))
print(len(d))
print(*d.values())

4
bcdef
abcdefg
bcde
bcdef
3
2 1 1


## Collections.deque()



In [61]:
from collections import deque

d = deque()
for _ in range(int(input())):
    attr, *kw = input().split()
    if kw:
        getattr(d,attr)(*(map(int, *kw)))
    else:
        getattr(d,attr)()
print(*d)

6
append 1
append 2
append 3
appendleft 4
pop
popleft
1 2


In [62]:
from collections import deque
d = deque()
for _ in range(int(input())):
    inp = input().split()
    getattr(d, inp[0])(*[inp[1]] if len(inp) > 1 else [])
print(*[item for item in d])

6
append 1
append 2
append 3
appendleft 4
pop
popleft
1 2


## Piling Up!

In [73]:
for t in range(int(input())):
    input()
    lst = list(map(int, input().split()))
    l = len(lst)
    i = 0
    while i < l - 1 and lst[i] >= lst[i+1]:
        i += 1
    while i < l - 1 and lst[i] <= lst[i+1]:
        i += 1
    print ("Yes" if i == l - 1 else "No")

2
6
4 3 2 1 3 4
Yes
3
1 3 2
No


In [71]:
from collections import deque
def check(d):
    while d:
        big = d.popleft() if d[0]>d[-1] else d.pop()
        print(big)
        if not d:
            return "Yes"
        if d[-1]>big or d[0]>big:
            return "No"
    
for i in range(int(input())):
    int(input())
    d = deque(map(int,input().split()))
    print(check(d))

2
6
4 3 2 1 3 4
4
4
3
3
2
1
Yes
3
1 3 2
2
No


In [69]:
d

deque([1, 3])

In [72]:
from collections import deque
def check(d):
    while d:
        big = d.popleft() if d[0]>d[-1] else d.pop()
#         print(big)
        if not d:
            return "Yes"
        if d[-1]>big or d[0]>big:
            return "No"
    
for i in range(int(input())):
    int(input())
    d = deque(map(int,input().split()))
    print(check(d))

2
6
4 3 2 1 3 4
Yes
3
1 3 2
No


## Company Logo

In [ ]:
#!/bin/python3

import math
import os
import random
import re
import sys
from collections import OrderedDict

if __name__ == '__main__':
    d = OrderedDict()
    s = input()
    

In [74]:
from collections import Counter, OrderedDict

class OrderedCounter(Counter, OrderedDict):
    pass
[print(*c) for c in OrderedCounter(sorted(input())).most_common(3)]

aabbbccde
b 3
a 2
c 2


[None, None, None]

In [76]:
from collections import Counter

class OrderedCounter(Counter):
    pass

[print(*c) for c in OrderedCounter(sorted(input())).most_common(3)]

aabbbccde
b 3
a 2
c 2


[None, None, None]

In [75]:
from collections import Counter

chars = Counter(input()).items()

for char, n in sorted(chars, key=lambda c: (-c[1], c[0]))[:3]:
    print(char, n)

aabbbccde
b 3
a 2
c 2
